qutebrowser/browser/bookmarks.py

```diff
@@ -0,0 +1,140 @@
+        _lineparser: The LineParser used for the bookmarks, or None
+                     (when qutebrowser is started with -c '').
+
+    Signals:
+        changed: Emitted when anything changed.
+        added: Emitted when a new bookmark was added.
+               arg 0: The title of the bookmark.
+               arg 1: The URL of the bookmark, as string.
+        removed: Emitted when an existing bookmark was removed.
+                 arg 0: The title of the bookmark.
+    """
+
+    changed = pyqtSignal()
+    added = pyqtSignal(str, str)
+    removed = pyqtSignal(str)
+
+    def __init__(self, parent=None):
+        """Initialize and read bookmarks."""
+        super().__init__(parent)
+
+        self.bookmarks = collections.OrderedDict()
+
+        if standarddir.config() is None:
+            self._lineparser = None
+        else:
+            bookmarks_directory = os.path.join(standarddir.config(),
+                                               'bookmarks')
+            if not os.path.isdir(bookmarks_directory):
+                os.makedirs(bookmarks_directory)
+            self._lineparser = lineparser.LineParser(
+                standarddir.config(), 'bookmarks/urls', parent=self)
+            for line in self._lineparser:
+                if not line.strip():
+                    # Ignore empty or whitespace-only lines.
+                    continue
+                try:
+                    url, title = line.split(maxsplit=1)
+                except ValueError:
+                    message.error(0, "Invalid bookmark '{}'".format(line))
+                else:
+                    self.bookmarks[url] = title
+            filename = os.path.join(standarddir.config(), 'bookmarks/urls')
+            objreg.get('save-manager').add_saveable(
+                'bookmark-manager', self.save, self.changed,
+                filename=filename)
+
+    def save(self):
+        """Save the bookmarks to disk."""
+        if self._lineparser is not None:
+            self._lineparser.data = [' '.join(tpl)
+                                     for tpl in self.bookmarks.items()]
+            self._lineparser.save()
+
+    def bookmark_add(self, win_id, url, title):
+        """Add a new bookmark.
+
+        Args:
+            win_id: The window ID to display the errors in.
+            url: The url to add as bookmark.
+            title: The title for the new bookmark.
+        """
+        if not url.isValid():
+            urlutils.invalid_url_error(win_id, url, "save quickmark")
+            return
+        urlstr = url.toString(QUrl.RemovePassword | QUrl.FullyEncoded)
+
+        # We don't raise cmdexc.CommandError here as this can be called async
+        # via prompt_save.
+        if not title:
+            message.error(win_id, "Can't set mark with empty title!")
+            return
+        if not urlstr:
+            message.error(win_id, "Can't set mark with empty URL!")
+            return
```

Comment:
Seeing this is called via `bookmark_add` in `commands.py`, shouldn't those checks always be true?